sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution
+
+import org.apache.spark.sql.catalyst.expressions.SortOrder
+import org.apache.spark.sql.catalyst.expressions.aggregate.{Complete, Final, Partial}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.aggregate.HashAggregateExec
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * Replace [[HashAggregateExec]] with [[SortAggregateExec]] in the spark plan if:
+ *
+ * 1. The plan is a pair of partial and final [[HashAggregateExec]], and the child of partial
+ *    aggregate satisfies the sort order of corresponding [[SortAggregateExec]].
+ * or
+ * 2. The plan is a [[HashAggregateExec]], and the child satisfies the sort order of
+ *    corresponding [[SortAggregateExec]].
+ *
+ * Examples:
+ * 1. aggregate after join:
+ *
+ *  HashAggregate(t1.i, SUM, final)
+ *               |                         SortAggregate(t1.i, SUM, complete)
+ * HashAggregate(t1.i, SUM, partial)   =>                |
```

Comment:
This seems like an orthogonal optimization: we can merge adjacent partial and final aggregates (no shuffle between them) into one complete aggregate.

Reply:
Yeah I think we can add a rule later to optimize it. I vaguely remember someone proposed this in OSS before but seems impact is not high.